x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtUtil.java

```diff
@@ -338,4 +348,35 @@ public static byte[] sha256(final CharSequence charSequence) {
         messageDigest.update(charSequence.toString().getBytes(StandardCharsets.UTF_8));
         return messageDigest.digest();
     }
+
+    /**
+     * Helper class to consolidate multiple trace level statements to a single trace statement with lazy evaluation.
+     * If trace level is not enabled, then no work is performed. This class is not threadsafe.
+     */
+    public static class TraceBuffer {
+        private final Logger logger;
+        List<Object> params = new ArrayList<>();
+        StringBuilder builder = new StringBuilder();
+
+        public TraceBuffer(Logger logger) {
+            this.logger = logger;
+        }
+
+        public void append(String s, Object... args) {
+            if (logger.isTraceEnabled()) {
+                builder.append(s).append(" ");
+                List<Object> resolved = Arrays.stream(args).map(x -> (x instanceof Supplier) ? ((Supplier<?>) x).get() : x).toList();
+                params.addAll(resolved);
+            }
+        }
+
+        @SuppressLoggerChecks(reason = "builds the tracer dynamically")
+        public void flush() {
+            if (logger.isTraceEnabled()) {
+                logger.trace(builder.toString(), params.toArray());
+                params = new ArrayList<>();
+                builder = new StringBuilder();
+            }
+        }
```

Comment:
I think it might be useful to make this class more stateful such that:
1. If `append` is never called, `flush` should be a noop.
2. Once it's flushed, it is closed and cannot take `append` or `flush` again.
3. Or if we want to reset with `flush`, we can call `params.clear()` and also replace `StringBuilder` with a `List<String>` so that it can `clear` as well. This helps making these fields `final`.

Reply:
Thanks. I thought about AutoCloseable as well but wasn't sure it's worth your time to adjust the code again for a tracing log. It's great to see you have made the change 👍